apisix/plugins/zipkin.lua

```diff
@@ -163,9 +165,9 @@ function _M.rewrite(plugin_conf, ctx)
         -- pass the trace ids even the sample is rejected
         -- see https://github.com/openzipkin/b3-propagation#why-send-
         -- trace-ids-with-a-reject-sampling-decision
-        core.request.set_header(ctx, "x-b3-traceid", trace_id)
-        core.request.set_header(ctx, "x-b3-parentspanid", parent_span_id)
-        core.request.set_header(ctx, "x-b3-spanid", request_span_id)
+        core.request.set_header(ctx, "x-b3-traceid", trace_id or to_hex(rand_bytes(16)))
```

Comment:
I am confused. Why do we generate a trace ID when the sampling is rejected?

Reply:
After reading the OpenTracing material, I can understand the reason.

The current implementation of APISIX is incomplete. For example, we don't have the ability to generate a root span. We require the client to pass a trace ID & span ID. Maybe we should add the ability to generate trace ID/span ID if not given instead of just generating a trace ID when the sampling is rejected

Reply:
Yes, I am wrong about the root span.

After some investigation, I figure out that:
Currently, APISIX handles root span like this:

1. check if need to sample or not
2. if not, set some upstream header and go to the next plugin
3. generate root span

The correct behavior should be:
1. check if need to sample or not
2. generate root span
3. set some upstream header (as the sampling should only affect the tracing data report)

In step 2 of APISIX, the root span is missing so we have to create a fake span ID in this PR.

Maybe we need to do some bigger changes to solve this problem?

For example, update
https://github.com/apache/apisix/blob/490e0ba995fb0bfe4c1b22eec424f58ebef39446/apisix/plugins/zipkin/codec.lua#L101
https://github.com/apache/apisix/blob/490e0ba995fb0bfe4c1b22eec424f58ebef39446/apisix/plugins/zipkin/reporter.lua#L58
to make it work correctly.

Otherwise, the generated fake span in this PR doesn't have a real span associated with it.

Reply:
I'm sorry, I didn't get your point. @spacewander 

Why do we need to generate a real root span when no sampling is required and the client does not pass a trace ID & span ID? Where does this real root span need to be used when no sampling is required?  We do not need to report these spans to the zipkin server.

A fake trace ID & span ID are sufficient. Only `upstream headers` and `apisix's logs` need these ids

Reply:
There is no standard way to do sampling in zipkin. But by reading the code of Kong / Envoy / Traefik, they both choose the "correct steps" we described in https://github.com/apache/apisix/pull/8099#discussion_r999020747.

The benefit of this way is that we don't need to duplicate header set operations in both sampling and no sampling path, especially when we need to introduce more flags in the future. The fake span id solution is not so popular.